### PR TITLE
Add gpu diag/perf cli job

### PR DESCRIPTION
**Pull request by jfomhover**
# PR into Azure/azureml-examples

This PR adds a single-step cli command job that runs performance checks on a gpu cluster. This provides 2 environments for the user to run those tests. It also provides a bash script to run some typical checks and help user troubleshoot low-performing configurations.

The README provides clarity on the value of this and how to use it.

## Checklist

I have:

- [ ] read and followed the contributing guidelines

## Changes

- adds a cli command job to run diagnostics/performance measurements in gpu nodes

### CODE REVIEW

File: cli/jobs/single-step/gpu_perf/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
+# Run NCCL tests on gpu to check performance and configuration
```
Comment (daholste) [inline cli/jobs/single-step/gpu_perf/README.md:1]: this is an amazing file... on an amazing PR...

File: cli/jobs/single-step/gpu_perf/environments/azureml/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+# set env var to find nccl rdma plugins inside this container
+ENV LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:/usr/local/nccl-rdma-sharp-plugins/lib"
+
+##############################################################################
+# NCCL TESTS
+##############################################################################
+ENV NCCL_TESTS_TAG=v2.11.0
+
+# NOTE: adding sm_35, sm_37 to support K80, M60
```
Comment (daholste) [inline cli/jobs/single-step/gpu_perf/environments/azureml/Dockerfile:14]: thorough testing re supporting legacy

**Review by daholste — APPROVED** [review]
Wow... what a thorough, incredible PR

Also, what a find to improve bandwidth by over an order of magnitude...

Amazing work obviously.... Thank you

File: cli/jobs/single-step/gpu_perf/README.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+3. Modify `gpu_perf_job.yml` to use your new environment name/version.
+
+4. Run the job using `az ml job create`.
+
+## Set environment variables
+
+In `gpu_perf_job.yml` you'll find an environment variables section that you can leverage for testing your specific configuration.
+
+For examples please see:
+- specs of [UCX environment variables](https://rocmdocs.amd.com/en/latest/Remote_Device_Programming/UCXenv.html)
```
Comment (lostmygithubaccount) [inline cli/jobs/single-step/gpu_perf/README.md:66]: remove locale in URL if possible. also these types of links tend to break over time, would recommend not having them (or doing the work in this repo to validate links -- docs does this)

File: cli/jobs/single-step/gpu_perf/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+languages:
+- azurecli
+- shell
+- docker
+products:
+- azure-machine-learning
+description: This sample shows how to run standard health checks on a GPU cluster to test performance and correctness of distributed multinode GPU jobs. It helps with troubleshooting performance issues related to the environment and container.
+---
+
+# Run NCCL tests on GPU to check performance and configuration
```
Comment (lostmygithubaccount) [inline cli/jobs/single-step/gpu_perf/README.md:12]: in general this kinda lends itself to a "how-to" doc -- why not host the code snippets in here and write a doc there?

File: cli/jobs/single-step/gpu_perf/gpu_perf_job.yml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+type: command
+
+code: 
+  local_path: src
+
+command: /bin/bash ./gpu_perf.sh
+
+# to iterate quickly on your environment design, use inline
+environment: file:./environments/azureml/env.yml
```
Comment (lostmygithubaccount) [inline cli/jobs/single-step/gpu_perf/gpu_perf_job.yml:12]: is this "file:" thing needed?

File: cli/jobs/single-step/gpu_perf/gpu_perf_job.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+compute: azureml:gpu-cluster
+
+distribution:
+  type: mpi
+  process_count_per_instance: 8 # NOTE: set equal to number of gpus on the node
+
+resources:
+  instance_count: 2 # NOTE: to use multiple nodes
+
+experiment_name: nccl-test
```
Comment (lostmygithubaccount) [inline cli/jobs/single-step/gpu_perf/gpu_perf_job.yml:44]: there is a naming convention (+ order of the YAML) sections in the contributing guide, would be good to align

Comment (lostmygithubaccount) [reply cli/jobs/single-step/gpu_perf/gpu_perf_job.yml:44]: same for display name

**Review by lostmygithubaccount — APPROVED** [review]

File: cli/jobs/single-step/gpu_perf/gpu_perf_job.yml (hunk 6 above)
Comment (jfomhover) [reply cli/jobs/single-step/gpu_perf/gpu_perf_job.yml:44]: I checked it and the yml follows the guidance provided, but for environment_variables and display_name there's no guidance.

here's that's in the guidelines https://github.com/Azure/azureml-examples/blob/main/cli/CONTRIBUTING.md#jobs
code > command > inputs > environment > compute > experiment_name > description

File: cli/jobs/single-step/gpu_perf/README.md (hunk 4 above)
Comment (jfomhover) [reply cli/jobs/single-step/gpu_perf/README.md:12]: Can you point me in the right direction on where to write how-tos ?

File: cli/jobs/single-step/gpu_perf/README.md (hunk 3 above)
Comment (jfomhover) [reply cli/jobs/single-step/gpu_perf/README.md:66]: Good point. Sadly I checked the url and removing the locale doesn't work.

File: cli/jobs/single-step/gpu_perf/gpu_perf_job.yml (hunk 5 above)
Comment (jfomhover) [reply cli/jobs/single-step/gpu_perf/gpu_perf_job.yml]: I thought it was, removing it worked.

Comment (jfomhover) [reply cli/jobs/single-step/gpu_perf/gpu_perf_job.yml:12]: Rolled back: if I don't use file:./ the run_job.sh will fail, it looks like a papercut in the way we handle abs/rel file paths there.

File: cli/jobs/single-step/gpu_perf/README.md (hunk 4 above)
Comment (lostmygithubaccount) [reply cli/jobs/single-step/gpu_perf/README.md:12]: done internally

**Review by balapv — APPROVED** [review]

**Pull request merged**